File: src/code/python/sorting_algorithms/counting_sort.py
def counting_sort(arr: list) -> list:
    max_num = max(arr)
    min_num = min(arr)
    count_range = max_num - min_num + 1
    count = [0] * count_range
    output = [0] * len(arr)

    for num in arr:
        count[num - min_num] += 1

    for i in range(1, count_range):
        count[i] += count[i - 1]

    for num in arr[::-1]:
        output[count[num - min_num] - 1] = num
        count[num - min_num] -= 1

    return output

File: src/code/ruby/sorting_algorithms/counting_sort.rb

def counting_sort(arr)
  max_num     = arr.max
  min_num     = arr.min
  count_range = max_num - min_num + 1
  count       = Array.new(count_range, 0)
  output      = Array.new(arr.size, 0)

  arr.each do |num|
    count[num - min_num] += 1
  end

  (1...count_range).each do |i|
    count[i] += count[i - 1]
  end

  arr.reverse_each do |num|
    output[count[num - min_num] - 1] = num
    count[num - min_num] -= 1
  end

  output
end
